---
title: "GWAS"
author: "Juan Jimenez"
date: "05-19-2022"
output: pdf_document
---

````{r setup, echo=FALSE}
## Path Normalization
knitr::opts_chunk$set(echo = TRUE) 
knitr::opts_knit$set(root.dir=normalizePath('')) 
#library(tidyverse)
```

```{bash, eval=FALSE}
####first###
# Genotypic data manipulation
mkdir ldata
cp -r data/RiceDiversity_44K_Genotypes_PLINK ldata/

module load plink/1.90
# convert it to binary file
cd ldata/RiceDiversity_44K_Genotypes_PLINK
plink --file sativas413 --make-bed --out binary_sativas413
```

```{bash, eval=FALSE}
####second###
### Calculating MAF and missingness uning plink
plink -bfile binary_sativas413 --freq --missing --out sativas413

# copy results back to cache folder!
cd ../../
cp ldata/RiceDiversity_44K_Genotypes_PLINK/sativas413.frq cache/
cp ldata/RiceDiversity_44K_Genotypes_PLINK/sativas413.lmiss cache/
```

```{bash, eval=FALSE}
####third###
git add --all
git commit -m "GWAS frq and missingness"
git push
```

```{r, eval=FALSE}
####fourth in local###
#to visualize MAF and missing rate
# install.packages("data.table")
library("data.table")

maf <- fread("cache/sativas413.frq", header=TRUE)
lmiss <- fread("cache/sativas413.lmiss", header=TRUE)


pdf("graphs/maf_lmiss.pdf", width = 10, height=5)
par(mfrow=c(1,2))
hist(maf$MAF, breaks=50, col="#cdb79e", main="MAF (SNP = 36,901)", xlab="Minor Allele Freq")
#abline(v=0.01, lty=2, col="black", lwd=3)
abline(v=0.05, lty=2, col="red", lwd=3)

hist(lmiss$F_MISS, breaks=35, col="#cdb79e", main="Missingness (SNP = 36,901)", xlab="Missing Rate")
#abline(v=0.6, lty=2, col="red", lwd=3)
#abline(v=0.05, lty=2, col="red", lwd=3)
dev.off()
```

```{bash, eval=FALSE}
####FIVE###
#LD decay
module load plink
cd ldata/RiceDiversity_44K_Genotypes_PLINK/
plink -bfile binary_sativas413 --r2 --ld-window 100 --ld-window-kb 100 --ld-window-r2 0 --out binary_sativas413 
```


```{r, eval=FALSE}
#in local##
# Summarize LD decay rate IN r
library("data.table")
# cd back to your project home dir
df <- fread("ldata/RiceDiversity_44K_Genotypes_PLINK/binary_sativas413.ld", data.table=FALSE)

BINSIZE = 100
df$dist <- df$BP_B - df$BP_A
df$bin <- round(df$dist/BINSIZE, 0)

library(plyr)

df2 <- ddply(df, .(bin), summarise,
      meanr2 = mean(R2))

write.table(df2, "cache/ld_in_100bp_bin.csv", sep=",", row.names=FALSE, quote=FALSE)

### Plot LD decay results AND SYNC TH GITHUB
ld <- read.csv("cache/ld_in_100bp_bin.csv")

pdf("graphs/ld_decay.pdf", width=10, height=10)
plot(ld$bin*100, ld$meanr2, xlab="Physical distance (bp)", ylab="R2", main="LD decay rate in rice")
abline(h=0.3, col="red")
dev.off()

```

```{bash, eval=FALSE}
####six###
# Population structure using PCA
cd ldata/RiceDiversity_44K_Genotypes_PLINK
plink -bfile binary_sativas413 --pca 'header' --out sativas413 
cd ../../
cp ldata/RiceDiversity_44K_Genotypes_PLINK/sativas413.eigenvec cache/
```

```{r, eval=FALSE}
#local#
# PCA using PLINK
### Plotting the PCA results in R

pca <- read.table("cache/sativas413.eigenvec", header=TRUE)


pdf("graphs/PCA.pdf", width=10, height=10)

plot(pca$PC1, pca$PC2, xlab="PC1", ylab="PC2")
plot(pca$PC3, pca$PC4, xlab="PC3", ylab="PC4")
plot(pca$PC5, pca$PC6, xlab="PC5", ylab="PC6")
plot(pca$PC7, pca$PC8, xlab="PC7", ylab="PC8")
plot(pca$PC9, pca$PC10, xlab="PC9", ylab="PC10")

dev.off()
```


```{r, eval=FALSE}
#in local##
#making a plot in R for the PCA 1,2,3

library("scatterplot3d")

fsize=16
pdf("graphs/pca_3d.pdf", width=10, height=10)
scatterplot3d(pca[,3:5], pch = 16, cex.symbol=1.2, color="#00BFC4", main="Maize Diversity Panel", angle=40)
dev.off()
```

```{r, eval=FALSE}
####seven###
# GWAS using the `gemma` software package
### Fit the QK model

#Using R to fit the QK model in HCC
module load R
R
library("data.table")

ped <- fread("sativas413.ped", header=FALSE)
ped$V6 <- 1
fwrite(ped, "sativas413.ped", sep="\t", row.names=FALSE, col.names = FALSE, quote=FALSE)

fam <- fread("sativas413.fam", header=FALSE)
fam$V6 <- 1
fwrite(fam, "sativas413.fam", sep="\t", row.names=FALSE, col.names = FALSE, quote=FALSE)
```

--

```{bash, eval=FALSE}
####eight###
module load plink
plink --file sativas413 --make-bed --out binary_sativas413
# To calculate centered relatedness matrix (will take ~ 1 min): Creating the GK matrix
module load gemma
gemma -bfile binary_sativas413 -gk 1 -o binary_sativas413

#Checking the data
module load R
R
library("data.table")
k <- fread("output/binary_sativas413.cXX.txt", header=FALSE)
dim(k)
#it result in 413 413
#q()

```

```{r, eval=FALSE}
####NINE###
# GWAS using the gemma software package
### Q matrix
#Creating the Q matrix for PCA as co variate
# cd to ldata/RiceDiversity_44K_Genotypes_PLINK
module load R
R

pca <- read.table("sativas413.eigenvec", header=TRUE)
pca[,2] <- 1
write.table(pca[,2:5], "pc3.txt", sep="\t", row.names=FALSE, 
            quote=FALSE, col.names = FALSE)

#q() to close
```

```{r fig.height=5, fig.width=5, message=FALSE, warning=FALSE, eval=FALSE}
# Phenotypic data
##local###
pheno <- read.delim("http://ricediversity.org/data/sets/44kgwas/RiceDiversity_44K_Phenotypes_34traits_PLINK.txt", header=TRUE)

library(ggplot2)
#Here you can replace x= for the trait you want to visualize in GWAS

pdf("graphs/histogram_panicle.pdf", width=10, height=10)

ggplot(pheno, aes(x=Panicle.length)) + 
  geom_histogram(aes(y=..density..), bins=50, fill="#999999")+
  geom_density(alpha=.2, fill="#FF6666") +
  labs(title="Phenotype histogram plot",x="Plant Height", y = "Density")+
  theme_classic()

dev.off()
```

```{r, eval=FALSE}

####TEN###
#writing the pheno.txt to the genotypic data folder on HCC
module loar R
R


pheno <- read.delim("http://ricediversity.org/data/sets/44kgwas/RiceDiversity_44K_Phenotypes_34traits_PLINK.txt", header=TRUE)
write.table(pheno[, -1:-2], "pheno.txt", 
            sep="\t", row.names=FALSE, quote=FALSE, col.names = FALSE)
dim(pheno)
#IT SHOULD BE 413 38
#q() to close
```

```{bash, eval=FALSE}
####elevent###
# GWAS using the gemma software package
#here chose -n number of the column you have the trait you want to work with example 13 for panicle.length and -o name of new gwas file panicle.length
module load gemma

gemma -bfile binary_sativas413 -c pc3.txt -k output/binary_sativas413.cXX.txt -p pheno.txt -lmm 4 -n 13 -o Plant.length -miss 0.9 -r2 1 -hwe 0 -maf 0.05
#to save the GWAS file in the cache folder
cp output/Plant.length.assoc.txt ../../cache
```

```{r, eval=FALSE}
####Tuelve###
# creating The Manhattan plot in desktop because library qqman is not in HCC
git add --all
git commit -m "GWAS RESULTS"
git push

library(qqman)
library("data.table")
res <- fread("cache/Plant.length.assoc.txt")

pdf("graphs/manhattan.pdf", width = 10, height=5)

manhattan(x = res, chr = "chr", bp = "ps", p = "p_wald", snp = "rs", col = c("blue4", "orange3"), logp = TRUE)

dev.off()
```


